File: code/cpp/easy/1337-kWeakestRows.cpp
#include <bits/stdc++.h>
#define PII pair<int, int>
using namespace std;

class compare {
   public:
    bool operator()(PII& a, PII& b) {
        if (a.second == b.second) {
            return a.first > b.first;
        } else {
            return a.second > b.second;
        }
    }
};

class Solution {
   public:
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        priority_queue<PII, vector<PII>, compare> pq;
        for (int i = 0; i < mat.size(); i++) {
            int soldiers = 0;
            for (int j = 0; j < mat[0].size(); j++) {
                if (mat[i][j] == 0) {
                    break;
                }
                soldiers += 1;
            }
            pq.push({i, soldiers});
        }

        vector<int> res;
        for (int i = 0; i < k; i++) {
            res.push_back(pq.top().first);
            pq.pop();
        }

        return res;
    }
};

File: code/java/easy/1337-kWeakestRows.java
class Pair implements Comparable<Pair> {
    int idx;
    int soldiers;

    Pair(int i, int sol) {
        soldiers = sol;
        idx = i;
    }

    @Override
    public int compareTo(Pair p2) {
        if (p2.soldiers == this.soldiers) {
            return this.idx - p2.idx;
        } else {
            return this.soldiers - p2.soldiers;
        }
    }
}

class Solution {
    public int[] kWeakestRows(int[][] mat, int k) {
        PriorityQueue<Pair> pq = new PriorityQueue<>();
        for (int i = 0; i < mat.length; i++) {
            int count = 0;
            for (int j = 0; j < mat[0].length; j++) {
                if (mat[i][j] == 0) {
                    break;
                }
                count += 1;
            }
            pq.add(new Pair(i, count));
        }

        int[] res = new int[k];
        for (int i = 0; i < k; i++) {
            res[i] = pq.remove().idx;
        }
        return res;
    }
}